### DISCUSSION TITLE
How to handle 64-bit memory addresses in the SCM commands from CLEO?

### DISCUSSION THREAD

**Discussion by fastman92** [cleolibrary/CLEO-Redux#34] [Ideas]
Newest GTA games have been compiled for 64-bit architectures.

For CLEO this has become a problem as far as memory commands are concerned.

I have thought about the best solution for this.

`

		enum class eArgumentDataTypesFormat_GTA_III_VC_SA : unsigned char
		{
			UNDEFINED,
			END_OF_ARGUMENTS = UNDEFINED,

			STATIC_INT_32BITS,			// Static number: (Long) Integer  – 32 bits
			GLOBAL_NUMBER_VARIABLE,		// Global 4-byte integer  /float variable multiplied by 4: (Short) Integer  – 16 bits
			LOCAL_NUMBER_VARIABLE,		// Local variable:  (Short) Integer  – 16 bits
			STATIC_INT_8BITS,			// Static number: (Char) Integer  – 8 bits
			STATIC_INT_16BITS,			// Static number: (Short) Integer  – 16 bits
			STATIC_FLOAT,				// Float IEEE 754 – 32 bits

			// Types below are only available in GTA SA
			TYPES_INTRODUCED_IN_GTASA,

			// Number arrays
			GLOBAL_NUMBER_ARRAY = TYPES_INTRODUCED_IN_GTASA,
			LOCAL_NUMBER_ARRAY,

			STATIC_SHORT_STRING,
			GLOBAL_SHORT_STRING_VARIABLE,
			LOCAL_SHORT_STRING_VARIABLE,

			GLOBAL_SHORT_STRING_ARRAY,
			LOCAL_SHORT_STRING_ARRAY,

			STATIC_PASCAL_STRING,
			STATIC_LONG_STRING,
			GLOBAL_LONG_STRING_VARIABLE,
			LOCAL_LONG_STRING_VARIABLE,

			GLOBAL_LONG_STRING_ARRAY,
			LOCAL_LONG_STRING_ARRAY,	// ID 19

			// added by Sanny Builder and CLEO
			STATIC_INT_64BITS,			// 20, Static number: (Long) Integer  – 64 bits, for example 0x12345678912357
			LOCAL_INT64_VARIABLE¨,	// 21, Local variable:  (Short) Integer  – 16 bits, for example 4@int64 in Sanny Builder`

Notice in the end two types have been added.

int64 type would use two variables internally. For example 4@int64 makes use of 4@ and 5@

These types will only be usable with few SCM commands - memory commands to mention, where memory address is expected.

// Reads one int64 value and updates the script pointer
int64_t CRunningScript::CollectInt64value();

CRunningScript::CollectInt64value would have to be implemented in CLEO and used by these few commands.

CRunningScript::CollectInt64value should check for STATIC_INT_64BITS / LOCAL_INT64_VARIABLE and read the value properly or call this->CollectParameters(1) to collect one number in standard way.

**Comment by x87** [comment]
What about WriteInt64value ?
```
4@int64 =  read_memory 1@int64 size 8 vp 0
```

what about offsets?
```
4@int64 += 1888
4@int64 =  read_memory 4@int64 size 8 vp 0
```

**Comment by fastman92** [comment]
Yes, implement WriteInt64 too. Check for STATIC_INT_64BITS / LOCAL_INT64_VARIABLE and write the value properly or call this->StoreParameters(1) to store one number in standard way.

"what about offsets?"

That's a good question with an easy solution.

As you see correctly, you also need certain commands to add/subtract the 64-bit values properly.

  `

    0A8E: get_sum_of_two_int_numbers 5@ + 6@ = 4@
    
    0A8F: get_difference_of_two_int_numbers 1@(0@,2i) - 1 = 4@
    
    0A90: get_multiplication_of_two_int_numbers 100 * 1@ = 30@
    
    0A91: get_division_of_two_int_numbers 8 / 2 = 0@
`

No need to create new commands. You may alter these ones.

Basically only certain CLEO SCM commands would be able to handle int64 values - commands for memory and simple mathematical operations.
